### PR TITLE
Added GridEditor to interactively define boundaries and grids

### PR DESCRIPTION
**Pull request by jlstevens**
This PR adds the `GridEditor` class which now incorporates much of the recent feedback:

## Core features

* Add, move and delete nodes. First node indicated with a special marker (triangle by default)

</img>

* Toggle node polarity (beta) with the Tap Tool. Generate Mesh button indicates when mesh generation is possible. Hide button to hide current grid/mesh.

</img>

* Easy insertion of new nodes into edges selected with the Tap Tool (which can then be easily moved):

</img>

* Pythonic access of boundary as it is drawn and the grid once it is generated:

</img>

* Easy adjustment of node size and edge width via the GUI. Easy regeneration with different x- and y-resolutions:

</img>

* Easy selection between a predefined set of tile sources in the background:

</img>

* Easy use of a custom element (e.g a different tile source) as the background:

</img>

* Download boundary as GeoJSON:

</img>

* Capture serializable editor state and restore from it (can also be restored from a geopandas boundary DataFrame):

</img>

* Set a focus function and update mesh accordingly:

</img>

* Highly customizable styling of boundary, nodes and start marker:

</img>

* Can be served as a panel dashboard using the `serveable()` method.

Note that I removed the `closed` button as 1) keeping the boundary open is actually a very helpful indicator of what is going on 2) the boundary is always closed implicitly by definition 3) no one seemed to really need this behavior 4) given the previous points it wasn't worth the complexity and potential confusion it added to the code.

## TODO

- [x] Clean up and commit notebook demonstrating usage
- [X] Wrap grid generation in try/except for the rare instances where the shape returned isn't what is expected. (issue was something else but this might be still needed if grid generation is found to fail)
- [x] Fix issue where initializing background to None causes problems.

## Wishlist

These are improvements I would like to make though I may not have time to do them immediately (and some of these are general OSS improvement suggestions)

- [ ] Support insertion of nodes into multiple selected edges (should be fairly easy)
- [ ] Select both new edges on either side of node edge insertions.
- [ ] Maybe we would like some way of defining the neutral boundary edges with bendy splines.
- [ ] Group the GUI elements nicely - I will file an issue on panel suggesting accordions based on parameter precedence ranges.

### CODE REVIEW

**Comment by phobson** [comment]
@jlstevens This is shaping up so well. Kudos.

**Comment by jlstevens** [comment]
Here is the panel issue https://github.com/holoviz/panel/issues/1632 that describes the mechanism I would like to implement my last wishlist item ('Group the GUI elements nicely')

**Comment by jlstevens** [comment]
Now that panel, param and holoviews have new releases on conda-forge, I've now built a conda package for `hologrid` which I've made available on my personal channel until it becomes a conda-forge feedstock as well:

```
conda install -c conda-forge -c jlstevens hologrid
```

I've tested this with Python 3.7 and 3.8 though I note that it takes a lot longer to solve for 3.8 than 3.7.

**Comment by jlstevens** [comment]
I have now address to TODO items above ignoring the wishlist items which I'll tick off once they are turned into issues instead. The documentation notebook has now been added and I've implemented various improvements suggested by Jim.

Note that this PR now removes the notebook with the initial spline prototype: the last commit with that notebook is 57f8302.

@jbednar I think this is ready to merge at which point we could implement a few more minor improvements or cut a new release.

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+from setuptools import setup
+
+required = {'param':">=1.10.0",
+            'panel':">=0.10.0",
+            'pygridgen':">0.2",
+            'geopandas':">=0.8.1",
+            'holoviews':">=1.0.1"}
```
Comment (jbednar) [inline setup.py]: Shouldn't these match what's in meta.yaml?

**Review by jbednar — APPROVED** [review]
I've pushed a few minor fixups, but otherwise it looks great to me!  Ready to merge once you answer the question about why the pip dependencies differ from the conda ones. I assume that's just been overlooked, but didn't want to change it if there was some reason.

File: setup.py (hunk 1 above)
Comment (jlstevens) [reply setup.py]: Yes it should. Fixed the holoviews pinning in 4947b56 (I checked the other pins and they are fine).

**Pull request merged**